### PR TITLE
coll/ucc: add support for gather(v), scatterv, reduce_scatter

### PR DESCRIPTION
**Pull request by Sergei-Lebedev**
Add support for more collectives in coll ucc: gather, gatherv, reduce_scatter, reduce_scatter_block, scatterv
Signed-off-by: Sergey Lebedev <sergeyle@nvidia.com>

### CODE REVIEW

**Review by bureddy — APPROVED** [review]

**Comment by edgargabriel** [comment]
a small question just for clarification: it seems like this commit adds support for scatterv, but not for scatter, is this in purpose? (the commit does add however support for gather and gatherv)

File: ompi/mca/coll/ucc/coll_ucc_gatherv.c [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    *request = &coll_req->super;
+    return OMPI_SUCCESS;
+fallback:
+    UCC_VERBOSE(3, "running fallback igatherv");
+    if (coll_req) {
+        mca_coll_ucc_req_free((ompi_request_t **)&coll_req);
+    }
+    return ucc_module->previous_igatherv(sbuf, scount, sdtype, rbuf, rcounts,
+                                         disps, rdtype, root, comm, request,
+                                         ucc_module->previous_igather_module);
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_gatherv.c:120]: previous_igather**V**_module

File: ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+                                              ucc_coll_req_h *req,
+                                              mca_coll_ucc_req_t *coll_req)
+{
+    ucc_datatype_t         ucc_dt;
+    ucc_reduction_op_t     ucc_op;
+    int total_count, i;
+    int comm_size = ompi_comm_size(ucc_module->comm);
+
+    if (MPI_IN_PLACE == sbuf) {
+        /* TODO: UCC defines inplace differently:
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c:25]: UCC_VERBOSE(5, "fallback ... ") ?

File: ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+static inline
+ucc_status_t mca_coll_ucc_reduce_scatter_init(const void *sbuf, void *rbuf, const int *rcounts,
+                                              struct ompi_datatype_t *dtype,
+                                              struct ompi_op_t *op, mca_coll_ucc_module_t *ucc_module,
+                                              ucc_coll_req_h *req,
+                                              mca_coll_ucc_req_t *coll_req)
+{
+    ucc_datatype_t         ucc_dt;
+    ucc_reduction_op_t     ucc_op;
+    int total_count, i;
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce_scatter.c:21]: make total_count size_t

File: ompi/mca/coll/ucc/coll_ucc_reduce_scatter_block.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+        UCC_VERBOSE(5, "ompi_op is not supported: op = %s",
+                    op->o_name);
+        goto fallback;
+    }
+    ucc_coll_args_t coll = {
+        .mask      = 0,
+        .coll_type = UCC_COLL_TYPE_REDUCE_SCATTER,
+        .src.info = {
+            .buffer   = (void*)sbuf,
+            .count    = rcount * comm_size,
```
Comment (vspetrov) [inline ompi/mca/coll/ucc/coll_ucc_reduce_scatter_block.c:47]: rcount * comm_size -> ((size_t)rcount) * comm_size to avoid potential INT overrun

**Comment by Sergei-Lebedev** [comment]
> a small question just for clarification: it seems like this commit adds support for scatterv, but not for scatter, is this on purpose? (the commit does add e.g. support for gather and gatherv)

yes, in UCC we don't have generic scatter implementation so far. Only device specific TLs (NCCL, RCCL, CUDA) supports it.

**Review by vspetrov — CHANGES_REQUESTED** [review]

**Comment by devreal** [comment]
>> a small question just for clarification: it seems like this commit adds support for scatterv, but not for scatter, is this on purpose? (the commit does add e.g. support for gather and gatherv)
>
> yes, in UCC we don't have generic scatter implementation so far. Only device specific TLs (NCCL, RCCL, CUDA) supports it.

Scatter is a specialization of scatterv. Why not implement the former in terms of the latter?

**Review by vspetrov — APPROVED** [review]

**Comment by janjust** [comment]
We're holding off until scater is added as a commit, then I'll merge

**Comment by Sergei-Lebedev** [comment]
@edgargabriel @devreal added scatter in last commit

**Comment by edgargabriel** [comment]
> @edgargabriel @devreal added scatter in last commit

excellent, thank you very much!

**Pull request merged**